### PR TITLE
Helm chart refactoring and improvements  

### LINKED ISSUE #263: Multiple namespace for openfaas within same kubernetes cluster
I am trying to run multiple namespace of openfaas within same kubernetes cluster but i am not able to install using helm

When i try to run with different namespace i am getting the below error
Error: release openfass-2 failed: clusterroles.rbac.authorization.k8s.io "faas-controller" already exists 

Is it possible to run openfaas with different namespace within same cluster and how can i achieve that?

### PR DESCRIPTION
**Pull request by stefanprodan**
This PR does the following:

- validate chart code on pull requests by running `helm lint` in TravisCI
- remove external services for Prometheus and Alertmanager #260
- remove internal and external services for faas-netes, the controller should be accessed only by the Gateway via localhost same as the operator
- add readiness and liveness probes to the Gateway, Prometheus and Alertmanager containers 
- add nodeSelector, tolerations and affinity to all components to enable node pool pinning 
- use Role instead of ClusterRole for the Operator, remove the Operator ClusterRole
- prefix the Operator Account, Role and RoleBinding with the Helm release name
- make the CRD creation optional (prep for multi-namespace support)
- prefix the faas-netes Account, Role and RoleBinding with the Helm release name
- make the Gateway nodePort configurable (must be changed with each install)
- remove NATS ClusterIP duplicate service
- add multiple release support for the same cluster, tested on GKE 1.10 with 4 instances (2x faas-netes and 2x operator) fix #263 #211
- update Gateway to v0.8.9 and openfaas-operator 0.8.2

## Motivation and Context

- [x] I have raised an issue to propose this change ([required](https://github.com/openfaas/faas/blob/master/CONTRIBUTING.md))

## How Has This Been Tested?
Tested on GKE with node selector.

Install command:

```
helm upgrade openfaas --install ./chart/openfaas \
    --namespace openfaas  \
    --set functionNamespace=openfaas-fn \
    --set serviceType=LoadBalancer \
    --set basic_auth=true \
    --set operator.create=true \
    --set nodeSelector."cloud\.google\.com/gke-nodepool"=default-pool
```

I have multiple node pools and OpenFaaS components were scheduled only on the default pool:

```
kubectl -n openfaas get pods -owide
NAME                            READY     STATUS    RESTARTS   AGE       IP           NODE
alertmanager-846454577d-9f5dw   1/1       Running   0          33s       10.28.1.11   gke-europe-default-pool-5b1f046e-9drs
gateway-7976b8b5d5-m69s7        2/2       Running   1          33s       10.28.0.9    gke-europe-default-pool-7948f671-gbvm
nats-f4c957464-c4xds            1/1       Running   0          33s       10.28.1.12   gke-europe-default-pool-5b1f046e-9drs
prometheus-6cb77fdcbf-hk89l     1/1       Running   0          33s       10.28.0.8    gke-europe-default-pool-7948f671-gbvm
queue-worker-789bf865dd-sn6np   1/1       Running   0          33s       10.28.0.10   gke-europe-default-pool-7948f671-gbvm
```

Only the Gateway gets an LB:

```
kubectl -n openfaas get svc
NAME               TYPE           CLUSTER-IP      EXTERNAL-IP     PORT(S)          AGE
alertmanager       ClusterIP      10.31.250.240   <none>          9093/TCP         8m
gateway            ClusterIP      10.31.242.76    <none>          8080/TCP         8m
gateway-external   LoadBalancer   10.31.243.112   35.234.101.46   8080:32121/TCP   8m
nats               ClusterIP      10.31.240.209   <none>          4222/TCP         8m
nats-external      ClusterIP      10.31.255.200   <none>          4222/TCP         8m
prometheus         ClusterIP      10.31.240.126   <none>          9090/TCP         8m
```

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I've read the [CONTRIBUTION](https://github.com/openfaas/faas/blob/master/CONTRIBUTING.md) guide
- [x] I have signed-off my commits with `git commit -s`
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

**Comment by stefanprodan** [comment]
@alexellis I've also tested the multi-release feature on baremetal on Kubernetes 1.11. This is ready for merge from my part.

File: chart/openfaas/templates/alertmanager-dep.yaml [hunk 1]
```diff
@@ -26,8 +26,18 @@ spec:
           - "alertmanager"
           - "--config.file=/alertmanager.yml"
           - "--storage.path=/alertmanager"
+        livenessProbe:
+          httpGet:
+            path: /#/status
+            port: 9093
+          timeoutSeconds: 30
```
Comment (alexellis) [inline chart/openfaas/templates/alertmanager-dep.yaml]: Curious where 30 seconds comes from? Could AlertManager take long to return? 

The value for `periodSeconds` is `10` in the [docs](https://kubernetes.io/docs/tasks/configure-pod-container/configure-liveness-readiness-probes/)

File: chart/openfaas/templates/controller-rbac.yaml [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+    chart: {{ .Chart.Name }}-{{ .Chart.Version }}
+    component: faas-controller
+    heritage: {{ .Release.Service }}
+    release: {{ .Release.Name }}
+  name: {{ .Release.Name }}-controller
+  namespace: {{ .Release.Namespace | quote }}
+{{- if .Values.rbac }}
+---
+apiVersion: rbac.authorization.k8s.io/v1beta1
+kind: ClusterRole
```
Comment (alexellis) [inline chart/openfaas/templates/controller-rbac.yaml]: Is this the ClusterRole for faas-netes? We didn't need to use ClusterRole in the plain YAML so this should probably run with least privileges too. The work is was tested via #212 but not merged.

File: chart/openfaas/templates/prometheus-external-svc.yaml [hunk 3]
```diff
@@ -1,25 +0,0 @@
...
-  labels:
-    app: {{ template "openfaas.name" . }}
-    chart: {{ .Chart.Name }}-{{ .Chart.Version }}
-    component: prometheus
-    heritage: {{ .Release.Service }}
-    release: {{ .Release.Name }}
-  name: prometheus-external
-  namespace: {{ .Release.Namespace | quote }}
-spec:
-  type: {{ .Values.serviceType }}
```
Comment (alexellis) [inline chart/openfaas/templates/prometheus-external-svc.yaml]: I think this should remain in the chart - if someone requests `"NodePort" .Values.serviceType` then the NodePort for Prometheus should be exposed. This is especially useful for development and bare-metal Kubernetes.

Comment (stefanprodan) [reply chart/openfaas/templates/prometheus-external-svc.yaml]: What about production on baremetal? You would expose Prom to everyone. With a baremetal cloud like Scaleway where you have no firewall this is very unsafe. I think the Helm chart should be secure by default, if you want to expose Prometheus for `dev` you can do `kubectl port-forward -n openfaas deployment/prometheus 9090:9090`

File: chart/openfaas/templates/alertmanager-dep.yaml (hunk 1 above)
Comment (stefanprodan) [reply chart/openfaas/templates/alertmanager-dep.yaml]: The value comes from the official chart https://github.com/helm/charts/tree/master/stable/prometheus also the Prom operator has the same values.

File: chart/openfaas/templates/prometheus-external-svc.yaml (hunk 3 above)
Comment (LucasRoesler) [reply chart/openfaas/templates/prometheus-external-svc.yaml]: I agree, using `port-forward` is pretty easy. We could add a comment in the readme or in the `NOTES.txt` to document this?

File: chart/openfaas/templates/controller-rbac.yaml (hunk 2 above)
Comment (stefanprodan) [reply chart/openfaas/templates/controller-rbac.yaml]: We can merge this as it is and ask for a rebase on #212.

File: chart/openfaas/templates/prometheus-external-svc.yaml (hunk 3 above)
Comment (alexellis) [reply chart/openfaas/templates/prometheus-external-svc.yaml]: The problem is that this is a breaking change and you haven't updated any docs and there are many tutorials that talk about this NodePort.

Comment (alexellis) [reply chart/openfaas/templates/prometheus-external-svc.yaml]: https://docs.openfaas.com/deployment/kubernetes/

Comment (alexellis) [reply chart/openfaas/templates/prometheus-external-svc.yaml]: Just the top few I found on Google:

https://medium.com/@IrekRomaniuk/openfaas-with-kubernetes-194e8eec48ab

http://panosgeorgiadis.com/blog/2017/11/08/how-to-start-with-openfaas/

https://github.com/justindavies/ServerlessOnAKS

https://blogs.technet.microsoft.com/livedevopsinjapan/2017/09/19/open-faas-on-acs-kubernetes/

https://hackernoon.com/deploying-kubernetes-on-premise-with-rke-and-deploying-openfaas-on-it-part-2-cc14004e7007

**Pull request merged**